src/rgw/driver/sfs/object.cc

```diff
@@ -282,26 +287,98 @@ void SFSObject::gen_rand_obj_instance_name() {
   state.obj.key.set_instance(buf);
 }
 
-int SFSObject::get_obj_attrs(
-  optional_yield y,
-  const DoutPrefixProvider *dpp,
-  rgw_obj *target_obj
-) {
-  lsfs_dout(dpp, 10) << " TODO" << dendl;
+/*
+  This should be intended as a mere fetch of object's attributes.
+  The caller will likely call get_attrs() after get_obj_attrs().
+  This is how we have interpreted this call.
+  The average usage of this seems to suggest that one first calls
+  get_obj_attrs() and then it uses get_attrs().
+  Doubts remain that this could be entirely correct for all cases;
+  so for the time being, we leave target_obj empty.
+  We rely on the fact that, if in the future something could potentially fail
+  because target_obj is left empty, that fail will be explicit.
+*/
+int SFSObject::get_obj_attrs(optional_yield y,
+                             const DoutPrefixProvider *dpp,
+                             rgw_obj *target_obj) {
+  refresh_meta();
+  return 0;
+}
+
+int SFSObject::get_obj_state(const DoutPrefixProvider *dpp,
+                             RGWObjState **_state,
+                             optional_yield y,
+                             bool follow_olh) {
+  *_state = &state;
   return 0;
 }
+
+int SFSObject::set_obj_attrs(const DoutPrefixProvider *dpp,
+                             Attrs *setattrs,
+                             Attrs *delattrs,
+                             optional_yield y) {
+  ceph_assert(objref);
+  auto meta = objref->meta;
+  map<string, bufferlist>::iterator iter;
+
+  if(delattrs) {
+    for(iter = delattrs->begin(); iter != delattrs->end(); ++iter) {
+      meta.attrs.erase(iter->first);
+    }
+  }
+  if(setattrs) {
+    for(iter = setattrs->begin(); iter != setattrs->end(); ++iter) {
+      meta.attrs[iter->first] = iter->second;
+    }
+  }
+
+  //synch attrs caches
+  state.attrset = attrs = meta.attrs;
+  state.has_attrs = true;
+
+  objref->metadata_flush_attrs(store);
+  return 0;
+}
+
+bool SFSObject::get_attr(std::string name, bufferlist& dest){
+  return objref->get_attr(name, dest);
+}
+
 int SFSObject::modify_obj_attrs(const char *attr_name,
                                        bufferlist &attr_val, optional_yield y,
                                        const DoutPrefixProvider *dpp) {
-  ldpp_dout(dpp, 10) << __func__ << ": TODO" << dendl;
-  return -ENOTSUP;
+  if(!attr_name){
+    return 0;
+  }
+  ceph_assert(objref);
+  auto meta = objref->meta;
+  meta.attrs[attr_name] = attr_val;
+
+  //synch attrs caches
+  state.attrset = attrs = meta.attrs;
+  state.has_attrs = true;
+
+  objref->metadata_flush_attrs(store);
+  return 0;
 }
 
 int SFSObject::delete_obj_attrs(const DoutPrefixProvider *dpp,
                                        const char *attr_name,
                                        optional_yield y) {
-  ldpp_dout(dpp, 10) << __func__ << ": TODO" << dendl;
-  return -ENOTSUP;
+  if(!attr_name){
```

Comment:
Why would the upper layer call this with an empty string attr name?

Reply:
I think I have copied this from another BE implementation; 

1.  does it make really sense to check `attr_name`? probably not.
2. it is really important in terms of computational cost? probably not.

Reply:
Is an empty string a special value perhaps? 

If it probably doesn't make sense to do this, I suggest finding that out.

Reply:
I honestly think this is not something worth to be addressed.
As above, feel free to investigate better if you believe.